src/tools/compiletest/src/load_cfg/itemlist.rs

```diff
@@ -0,0 +1,733 @@
+
+use std::{
+    cell::{Cell, LazyCell},
+    ops::ControlFlow,
+    rc::Rc,
+    sync::LazyLock,
+};
+
+use super::{CommentTy, Error, ErrorExt, LineCol, Result};
+use regex::Regex;
+
+const STRSIM_CONFIDENCE: f64 = 0.7;
+
+pub fn parse(s: &str, cty: CommentTy) -> Result<Vec<Item>, Vec<Error>> {
+    let mut items = Vec::new();
+    let mut errors = Vec::new();
+
+    for (idx, line) in s.lines().enumerate() {
+        let line_no = idx + 1;
+        match try_match_line(line, cty).line(line_no) {
+            Ok(Some(item)) => items.push(Item::new(item, line_no, 1)),
+            Ok(None) => (),
+            Err(e) => errors.push(e),
+        }
+    }
+
+    Item::visit_expand_nested(&mut items).map_err(|e| errors.push(e));
+
+    if errors.is_empty() { Ok(items) } else { Err(errors) }
+}
+
+/// A type for mapping `CommentTy`s to an associated regex.
+type CommentRePats = [(CommentTy, Regex); CommentTy::all().len()];
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Item<'src> {
+    pub val: ItemVal<'src>,
+    pub pos: LineCol,
+    /// Purely a debugging tool; we set this to `true` when we consume the `Item` in some way
+    /// when setting the config (other module), which allows us to assert that nothing
+    /// accidentally goes unused.
+    pub(super) used: Cell<bool>,
+}
+
+impl<'src> Item<'src> {
+    fn new(item: ItemVal<'src>, line: usize, col: usize) -> Self {
+        Self { val: item, pos: LineCol { line, col }, used: Cell::new(false) }
+    }
+
+    /// Expand any items that may have nested items.
+    fn visit_expand_nested(list: &mut Vec<Item>) -> Result<()> {
+        for item in list.iter_mut() {
+            match item.val {
+                // Revision-specific items need to have their content parsed. E.g.
+                // `//@ [abc,def] compile-flags: -O`
+                ItemVal::RevisionSpecificItems { revs, content } => {
+                    let make_err = || {
+                        Err("unable to parse revision directive '{content}'".into()).pos(item.pos)
+                    };
+                    let content = match_any(
+                        |matcher, pass| matcher.try_match_directive(content, pass),
+                        make_err,
+                    )?
+                    .ok_or_else(|| make_err().unwrap_err())?;
+
+                    *item = Item {
+                        val: ItemVal::RevisionSpecificExpanded { revs, content: Box::new(content) },
+                        pos: item.pos,
+                        used: Cell::new(false),
+                    };
+                }
+                ItemVal::RevisionSpecificExpanded { .. } => unreachable!(
+                    "if this visit has never happened before, there should be \
+                    no expanded nodes."
+                ),
+                _ => (),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// TODO: parse revisions into `ItemVal`s.
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ItemVal<'src> {
```

Comment:
Discussion: so this is what I had in mind when I mentioned test suite specific directives: some of these directives make no sense for ui/ui-fulldeps but not for run-make, for example, and I think we might want to separate `ItemVal` (I would even just call this a `Directive`?) for specific test suites.

Reply:
`Directive` does make sense, my naming isn't as clean as it could be 😄. I think that having all directives listed in one place simplifies things, but I was planning to handle this. roughly:

```
trait TestKind {
    type Config;
    /// Called within `itemlist` to filter which directives we look for
    fn accepts_directive(m: &MatchItem /* I know, MatchItem is a bad name */) -> bool;

    /// Called within `prepare` to set its own config
    fn visit_build_config(item: &Item, cfg: &mut Self::Config) -> Result<()>;
}
```

This would be the only test-specific interface needed - just needs to say whether or not it supports a specific directive kind, and then say how to update its config when it gets to relevant items.

(Still mind mapping that bit out, but I think this seems reasonably clean and encapsulated)